Add scmi cpu support

ZhaoQiang-b45475 opened the pull request:
This PR add a driver NXP SCMI cpu dmomain protocol to set and get cpu status

ZhaoQiang-b45475 committed:
drivers: firmware: scmi: add cpu domain protocol

Added helpers for NXP SCMI cpu dmomain protocol.

Signed-off-by: Qiang Zhao <qiang.zhao@nxp.com>

AndreHeinemans-NXP commented on drivers/firmware/scmi/nxp/Kconfig:
```diff
@@ -6,3 +6,14 @@ config ARM_SCMI_NXP_VENDOR_EXTENSIONS
 	help
 	  When enabled, additional SCMI features specific to NXP will be available
+
+if ARM_SCMI
```
This whole file is sourced only if ARM_SCMI is enabled so this 'if' is not needed.

Consider making it dependent on ARM_SCMI_NXP_VENDOR_EXTENSIONS

ZhaoQiang-b45475 replied:
dependent on ARM_SCMI_NXP_VENDOR_EXTENSIONS is ok

PetervdPerk-NXP commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -53,6 +56,13 @@ static int soc_init(void)
+	cpu_cfg.cpu_id = CPU_IDX_M7P;
+	cpu_cfg.sleep_mode = CPU_SLEEP_MODE_RUN;
+
+	ret = scmi_cpu_sleep_mode_set(&cpu_cfg);
```
This only gets enabled when `CONFIG_ETH_NXP_IMX_NETC` is enabled. Please move outside the if statement.
Also you've to guard it with `CONFIG_NXP_SCMI_CPU_DOMAIN_HELPERS` otherwise compilation fails.

ZhaoQiang-b45475 replied:
Thanks for your commit, have fixed.

JiafeiPan commented on include/zephyr/drivers/firmware/scmi/cpu.h:
```diff
@@ -0,0 +1,64 @@
+	SCMI_CPU_DOMAIN_MSG_CPU_PD_LPM_CONFIG_SET = 0xA,
+	SCMI_CPU_DOMAIN_MSG_CPU_PER_LPM_CONFIG_SET = 0xB,
+	SCMI_CPU_DOMAIN_MSG_CPU_INFO_GET = 0xC,
+	SCMI_CPU_DOMAIN_MSG_NEGOTIATE_PROTOCOL_VERSION = 0x10,
```
Is these protocol NXP specific?  if yes, suggest rename the header file to be nxp_scmi_cpu.h or move it to a sub-directory of nxp

ZhaoQiang-b45475 replied:
send a new version to dix

JiafeiPan commented on include/zephyr/dt-bindings/cpu/imx95_cpu.h:
```diff
@@ -0,0 +1,31 @@
+/*! Select between GPC or GIC for wakeup */
+#define CPU_SLEEP_FLAG_IRQ_MUX    0x1U
+/*! Wake A55 CPU during A55 platform wakeup */
+#define CPU_SLEEP_FLAG_A55P_WAKE  0x2U
```
Are these definitions will be used by dts? if yes, I don't find any related description is dts binding yaml file, if not, suggest to move it to soc directory and rename it to be like "soc_scmi_cpu.h", and include it in nxp_scmi_cpu.h like
```
#if __has_include("soc_scmi_cpu.h")
#include "soc_scmi_cpu.h"
#endif
```

ZhaoQiang-b45475 replied:
send a new version to fix

ZhaoQiang-b45475 committed:
dts: arm: nxp_imx95_m7: add cpu domain node

Added cpu domain node on imx95 core m7

Signed-off-by: Qiang Zhao <qiang.zhao@nxp.com>

ZhaoQiang-b45475 committed:
soc: nxp: imx95: fix systick issue for M7

The SLEEP_HOLD_EN is enabled by default, that will
gate systick, clear it to fix

Signed-off-by: Qiang Zhao <qiang.zhao@nxp.com>

LaurentiuM1234 commented on doc/hardware/arch/arm-scmi.rst:
```diff
@@ -139,6 +139,8 @@ Currently, Zephyr has support for the following standard protocols:
 	#. **Clock management**
 	#. **Pin Control**
 
+NXP-specific protocols:
```
Maybe it would be good to have a separate documentation page for each vendor? Also, perhaps it would be useful for the community to document the vendor protocols? Like Linux did here: https://elixir.bootlin.com/linux/v6.13.7/source/drivers/firmware/arm_scmi/vendors/imx/imx95.rst.

CC: @kartben any takes on this?

ZhaoQiang-b45475 replied:
Hi Laurentiu, Thanks for your comments, they are very helpful. Will fix in next version, I also have some comments, would you please review?

LaurentiuM1234 commented on drivers/firmware/scmi/nxp/Kconfig:
```diff
@@ -6,3 +6,14 @@ config ARM_SCMI_NXP_VENDOR_EXTENSIONS
+
+if ARM_SCMI_NXP_VENDOR_EXTENSIONS
+
+config NXP_SCMI_CPU_DOMAIN_HELPERS
```
I think we need to get rid of this "*_HELPERS" naming scheme. It just doesn't make sense anymore for cases such as this one where there's no subsystem API....

anyways, I'd say leave as-is until we find a solution.

ZhaoQiang-b45475 replied:
So we could keep it this time and fix later

LaurentiuM1234 commented on dts/bindings/firmware/nxp,scmi-cpu.yaml:
```diff
@@ -0,0 +1,13 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: System Control and Management Interface (SCMI) cpu domain protocol
```
maybe I'm a bit paranoid but can you please add some comment in the description stating that this is a vendor-specific protocol? I know we already have the "nxp" prefix but I want to make sure there's no confusion.

LaurentiuM1234 commented on soc/nxp/imx/imx9/imx95/scmi_cpu_soc.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef ZEPHYR_NXP_IMX95_SCMI_CPU_SOC_H_
+#define ZEPHYR_NXP_IMX95_SCMI_CPU_SOC_H_
+
+#define CPU_IDX_M33P        0U
```
which of these macros are soc-specific and protocol-specific? if there's any protocol-specific stuff (i.e. that won't change across SoCs) then you can just add it to the protocol header. Looking at this header my guess is the CPU indexes are SoC-specific while the CPU modes might be protocol specific?

either way, I'd say there's no need to have this header included in the protocol header. You can just explicitly include it wherever needed?

ZhaoQiang-b45475 replied:
The CPU indexes are SoC-specific. I am not sure if the CPU modes are SoC-specific, I copy them from System Manager which put them into a soc header.

ZhaoQiang-b45475 replied:
Hi Laurentiu,
Thanks for your comments, the struct scmi_cpu_sleep_mode_config is defined in protocol header,
![image](https://github.com/user-attachments/assets/effbb058-f728-433f-a376-0ef7e3634a54)
The macros in **soc.h is the value of the struct, the scmi_cpu protocol depends on the macros, so I think it is better to include the soc.h in protocol header.

LaurentiuM1234 commented on soc/nxp/imx/imx9/imx95/m7/soc.c:
```diff
@@ -69,7 +74,18 @@ static int soc_init(void)
+	cpu_cfg.cpu_id = CPU_IDX_M7P;
+	cpu_cfg.sleep_mode = CPU_SLEEP_MODE_RUN;
+
+	ret = scmi_cpu_sleep_mode_set(&cpu_cfg);
```
can you please provide a more detailed commit description on this? it is very important to understand why we need this change as it will affect multiple setups.
